pkg/scheduler/internal/cache/node_tree.go

```diff
@@ -135,36 +121,32 @@ func (nt *nodeTree) updateNode(old, new *v1.Node) {
 	nt.addNode(new)
 }
 
-func (nt *nodeTree) resetExhausted() {
-	for _, na := range nt.tree {
-		na.lastIndex = 0
-	}
-	nt.zoneIndex = 0
-}
-
-// next returns the name of the next node. NodeTree iterates over zones and in each zone iterates
+// listNodes returns the list of namesof the node. NodeTree iterates over zones and in each zone iterates
 // over nodes in a round robin fashion.
-func (nt *nodeTree) next() string {
+func (nt *nodeTree) list() ([]string, error) {
+	nodesList := make([]string, 0, nt.numNodes)
 	if len(nt.zones) == 0 {
-		return ""
+		return nodesList, nil
 	}
 	numExhaustedZones := 0
-	for {
-		if nt.zoneIndex >= len(nt.zones) {
-			nt.zoneIndex = 0
-		}
-		zone := nt.zones[nt.zoneIndex]
-		nt.zoneIndex++
-		// We do not check the exhausted zones before calling next() on the zone. This ensures
-		// that if more nodes are added to a zone after it is exhausted, we iterate over the new nodes.
-		nodeName, exhausted := nt.tree[zone].next()
-		if exhausted {
-			numExhaustedZones++
-			if numExhaustedZones >= len(nt.zones) { // all zones are exhausted. we should reset.
-				nt.resetExhausted()
+	allocatedNodes := 0
+	nodeIndex := 0
+	for allocatedNodes < nt.numNodes {
+		for zoneIndex := 0; zoneIndex < len(nt.zones); zoneIndex++ {
+			zone := nt.zones[zoneIndex]
+			if nodeIndex >= len(nt.tree[zone].nodes) { // If the zone is exhausted, continue
+				if nodeIndex == len(nt.tree[zone].nodes) { // If it is the first time the zone is exhausted
+					numExhaustedZones++
+				}
+				continue
 			}
-		} else {
-			return nodeName
+			nodesList = append(nodesList, nt.tree[zone].nodes[nodeIndex])
+			allocatedNodes++
+		}
+		nodeIndex++
+		if numExhaustedZones >= len(nt.zones) { // all zones are exhausted.
```

Comment:
given the implementation, I think it's possible this is true, when we have already reached `allocatedNodes = nt.numNodes`.

Maybe just move it to the beginning of the loop (below `for allocatedNodes < nt.NumNodes`)

Reply:
good catch, thank you

Reply:
UT didn't catch that?

Reply:
>UT didn't catch that? if not, we should add one.

ping on this.